### ISSUE TITLE
Compiling dftatom

### ISSUE THREAD

**Issue by certik** [lfortran/lfortran#2386]
We currently use this repository to test and apply workarounds: https://github.com/czgdp1807/dftatom

One can build the project using:
```console
dftatom $ git clean -fdx; make -f Makefile.manual       
```
One can run the tests using:
```console
dftatom $ make -f Makefile.manual test
```
  

ASR Pass issues
--
- [x] https://github.com/lfortran/lfortran/issues/2434 - https://github.com/lfortran/lfortran/pull/2448
- [x] https://github.com/lfortran/lfortran/issues/2435 - https://github.com/lfortran/lfortran/pull/2438
- [x] https://github.com/lfortran/lfortran/issues/2664
- [x] https://github.com/lfortran/lfortran/issues/2667
- [x] https://github.com/lfortran/lfortran/issues/2670

LLVM issues
--
- [x] https://github.com/lfortran/lfortran/issues/2439 - https://github.com/lfortran/lfortran/pull/2450 (needs to wait)
- [x] https://github.com/lfortran/lfortran/issues/2436 - https://github.com/lfortran/lfortran/pull/2441
- [x] https://github.com/lfortran/lfortran/issues/2437 - https://github.com/lfortran/lfortran/pull/2488
- [x] https://github.com/lfortran/lfortran/issues/2461
- [x] https://github.com/lfortran/lfortran/issues/2462
- [x] https://github.com/lfortran/lfortran/issues/2547
- [x] https://github.com/lfortran/lfortran/issues/2671

Runtime issues
--
- [x] https://github.com/lfortran/lfortran/issues/2472
- [x] https://github.com/lfortran/lfortran/issues/2491
- [x] https://github.com/lfortran/lfortran/issues/2554
- [x] https://github.com/lfortran/lfortran/issues/2562
- [x] https://github.com/lfortran/lfortran/issues/2561
- [x] https://github.com/lfortran/lfortran/issues/2601
- [x] https://github.com/lfortran/lfortran/issues/2618
- [x] https://github.com/lfortran/lfortran/issues/2619
- [x] https://github.com/lfortran/lfortran/issues/2620
- [x] https://github.com/lfortran/lfortran/issues/2621
- [x] https://github.com/lfortran/lfortran/issues/2622
- [x] https://github.com/lfortran/lfortran/issues/2627

Current workarounds
--

- [x] https://github.com/certik/dftatom/commit/c1b5ccaa592961d7a3acf6b87b3c5059e543bc87
- [x] https://github.com/czgdp1807/dftatom/commit/92c09f4f13cf06f7c82997a76d703af2d17d89aa - Fixed by https://github.com/lfortran/lfortran/pull/2448
- [x] https://github.com/czgdp1807/dftatom/commit/290c5127cab76547f115fcd7bccccc7bcd022786 - Fixed by https://github.com/lfortran/lfortran/pull/2448
- [x] https://github.com/czgdp1807/dftatom/commit/572a2d9849db2a079e2e07c2e6bf159c5e83dee1 - Fixed by https://github.com/lfortran/lfortran/pull/2438
- [x] https://github.com/czgdp1807/dftatom/commit/c1abff82a192bc669f23830aa77fe80ab70a25fe
- [x] https://github.com/czgdp1807/dftatom/commit/6f0270fc17bf195de6189963a5f428de4723d85b - Fixed by https://github.com/lfortran/lfortran/pull/2488
- [x] https://github.com/czgdp1807/dftatom/commit/4a80c047c507c61c9eed7ba4332ea64494317147 - Fixed by https://github.com/lfortran/lfortran/pull/2507
- [x] https://github.com/czgdp1807/dftatom/commit/6ead4bc47871ee83cdae6d39af93607599e105ae - Fixed by https://github.com/lfortran/lfortran/pull/2504
- [x] https://github.com/czgdp1807/dftatom/commit/b40572d195b74109dff25cf208e18b7cb8844aa5
- [x] https://github.com/czgdp1807/dftatom/commit/976ca21ce97da46041a9f8c9e5af3da53667cf81
- [x] https://github.com/czgdp1807/dftatom/commit/ab026702c8536e3fb84dabde9eec58f80bead1a3 Reverted here: https://github.com/czgdp1807/dftatom/commit/7c117fe03ee1562e6cc913bdc7807416fd7dc41a
- [x] https://github.com/czgdp1807/dftatom/commit/f3b87b1261984d3a2de583448c77da2b23e29ca0
- [x] https://github.com/czgdp1807/dftatom/commit/7e1e2f851e4a24f7f9d6d18e5a8aa23022b6749c
- [x] https://github.com/czgdp1807/dftatom/commit/454b6aaaa2ec38a0472c0455ecff6f38e0c32f69
- [x] https://github.com/czgdp1807/dftatom/commit/104770cb50b919ac9262fb0b5353a84c74f55e6e
- [x] https://github.com/czgdp1807/dftatom/commit/0cd4a473d96f137cac143415bf28a418af2873f6
- [x] https://github.com/czgdp1807/dftatom/commit/6eace72ebd20abc5e4e9d0d2f7aa31ebbc409c54
- [x] https://github.com/czgdp1807/dftatom/commit/cc25712d59ffaf6052af297a308a53e85f48d197
- [x] https://github.com/czgdp1807/dftatom/commit/5b23e139782042c5f39131d603e5e9764d8fe89b
- [x] https://github.com/czgdp1807/dftatom/commit/b70e61bfa11a37468b2dd598f109336889f849e1
- [x] https://github.com/czgdp1807/dftatom/commit/bbdf2b50228a368be786a46a04b279f2dd12bddb
- [x] https://github.com/czgdp1807/dftatom/commit/aca350f4652065f0da3b3607cdbd711c73e0c7af
- [x] https://github.com/czgdp1807/dftatom/commit/26ee7bf40db1eaa915e46003394f573457f7ba06
- [x] https://github.com/czgdp1807/dftatom/commit/a73ca412dce71542542b3c1ec3c128e44490e422

Printing issues:
--
- [x] https://github.com/lfortran/lfortran/issues/2642
- [x] https://github.com/lfortran/lfortran/issues/2643

Low priority
--
- [ ] https://github.com/lfortran/lfortran/issues/2421
- [ ] https://github.com/lfortran/lfortran/issues/2433
- [ ] https://github.com/czgdp1807/dftatom/commit/9fa9de14b2fc3fd9e15cab69ce23a993afbd5545
- [x] https://github.com/czgdp1807/dftatom/commit/33dec0128e6e42b2d9e4687fec63fcff5c8c85b3
- [ ] https://github.com/czgdp1807/dftatom/commit/fd3b4f6be20e4753953ece8ff258a45602f28bf7

Bonus
--
- [ ] https://github.com/lfortran/lfortran/issues/2706

**Comment by czgdp1807** [comment]
I will work on it. I will update here when can I work on it. Either from day after tomorrow or from 11th September.

**Comment by czgdp1807** [comment]
I will work on it next week after completing libasr sync and refactoring of array_op pass for functions returning arrays.

**Comment by Shaikh-Ubaid** [comment]
Minimum reproducible example:

```fortran
$ cat examples/expr2.f90     
module data
    implicit none
    public data_t

    type data_t
        integer :: x
    end type
end module

program main
    use data, only: data_t
    implicit none

    print *, get_x_from_data(return_x_from_data)
    contains

    function return_x_from_data(d)
        type(data_t), intent(inout) :: d
        integer :: return_x_from_data
        return_x_from_data = d%x
    end function

    function get_x_from_data(R)
    interface
        function R(d)
        use data
        implicit none
        type(data_t), intent(inout) :: d
        integer :: R
        end function
    end interface
    integer :: get_x_from_data
    type(data_t) :: d
    d%x = 100
    get_x_from_data = R(d)
    end function

end program
```
```bash
$ gfortran examples/expr2.f90
$ ./a.out 
         100
```
```bash
$ lfortran examples/expr2.f90
ASR verify pass error: ASR verify: The symbol table was not found in the scope of `symtab`.
 --> examples/expr2.f90:5:5 - 7:13
  |
5 |        type data_t
  |        ^^^^^^^^^^^...
...
  |
7 |        end type
  | ...^^^^^^^^^^^^^ failed here

ASR verify pass error: ASR verify: Struct::m_derived_type 'data_t' cannot point outside of its symbol table, owner: r
  --> examples/expr2.f90:28:9
   |
28 |         type(data_t), intent(inout) :: d
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ failed here


Note: Please report unclear or confusing messages as bugs at
https://github.com/lfortran/lfortran/issues.
```

**Comment by Shaikh-Ubaid** [comment]
This can probably be reduced further as follows:
```fortran
$ cat examples/expr2.f90 
module data
    implicit none
    public data_t

    type data_t
        integer :: x
    end type
end module

module main
    use data, only: data_t
    implicit none

    contains

    function get_x_from_data(R)
    interface
        function R(d)
        use data
        implicit none
        type(data_t), intent(inout) :: d
        integer :: R
        end function
    end interface
    type(data_t) :: d
    integer :: get_x_from_data
    get_x_from_data = R(d)
    end function

end module
```
```bash
$ gfortran -c examples/expr2.f90
```
```bash
$ lfortran -c examples/expr2.f90
ASR verify pass error: ASR verify: The symbol table was not found in the scope of `symtab`.
 --> examples/expr2.f90:5:5 - 7:13
  |
5 |        type data_t
  |        ^^^^^^^^^^^...
...
  |
7 |        end type
  | ...^^^^^^^^^^^^^ failed here

ASR verify pass error: ASR verify: Struct::m_derived_type 'data_t' cannot point outside of its symbol table, owner: r
  --> examples/expr2.f90:21:9
   |
21 |         type(data_t), intent(inout) :: d
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ failed here


Note: Please report unclear or confusing messages as bugs at
https://github.com/lfortran/lfortran/issues.
```

**Comment by Thirumalai-Shaktivel** [comment]
Current status of the branch: `lf2`
```console
$ make -f Makefile.manual               
cd src; make -f Makefile.manual
lfortran  -c types.f90
lfortran  -c constants.f90
lfortran  -c dft_data.f90
lfortran  -c utils.f90
lfortran  -c energies.f90
lfortran  -c ode1d.f90
lfortran  -c mixings.f90
lfortran  -c rdirac.f90
lfortran  -c rschroed.f90
lfortran  -c reigen.f90
lfortran  -c rpoisson.f90
lfortran  -c dft.f90
lfortran  -c mesh.f90
lfortran  -c states.f90
lfortran  -c drivers.f90
lfortran  -c dftatom.f90
ar rcs libdftatom.a constants.o dft.o dft_data.o dftatom.o drivers.o energies.o mesh.o mixings.o ode1d.o rdirac.o reigen.o rpoisson.o rschroed.o states.o types.o utils.o
warning: /Users/thirumalai/mambaforge/envs/fpm/bin/arm64-apple-darwin20.0.0-ranlib: archive library: libdftatom.a the table of contents is empty (no object file members in the library define global symbols)
cd tests; make -f Makefile.manual
cd lda; make -f Makefile.manual
lfortran  -I../../src -c conv_lda.f90
ASR verify pass error: ASR verify: The input and output type of elemental intrinsics must exactly match, input type: r64[:] output type: r64
Internal Compiler Error: Unhandled exception
Traceback (most recent call last):
  File "/Users/thirumalai/Open_Source/lfortran/src/bin/lfortran.cpp", line 2251
    return compile_to_object_file(arg_file, outfile, false,
  File "/Users/thirumalai/Open_Source/lfortran/src/bin/lfortran.cpp", line 962
    res = fe.get_llvm3(*asr, lpm, diagnostics, infile);
  File "/Users/thirumalai/Open_Source/lfortran/src/lfortran/fortran_evaluator.cpp", line 347
    compiler_options, run_fn, infile);
  File "/Users/thirumalai/Open_Source/lfortran/src/libasr/codegen/asr_to_llvm.cpp", line 8955
    pass_manager.apply_passes(al, &asr, pass_options, diagnostics);
  File "/Users/thirumalai/Open_Source/lfortran/src/libasr/pass/pass_manager.h", line 303
    apply_passes(al, asr, _passes, pass_options, diagnostics);
  File "/Users/thirumalai/Open_Source/lfortran/src/libasr/pass/pass_manager.h", line 168
    throw LCompilersException("Verify failed in the pass: "
LCompilersException: Verify failed in the pass: array_op
make[2]: *** [conv_lda.o] Error 1
make[1]: *** [all] Error 2
make: *** [all] Error 2

```

**Comment by czgdp1807** [comment]
Okay. Now that needs to be fixed if possible. Let’s investigate. @Thirumalai-Shaktivel Can you create an MRE for it? Once MRE is available we will see if either to workaround or to proceed after we do the fix.

**Comment by Thirumalai-Shaktivel** [comment]
Yup, working on it!

**Comment by czgdp1807** [comment]
Seems like some elemental function didn’t get processed correctly by the array_op ASR pass.

**Comment by Thirumalai-Shaktivel** [comment]
Current issue:
```console
$ make -f Makefile.manual
cd src; make -f Makefile.manual
lfortran  -c types.f90
lfortran  -c constants.f90
lfortran  -c dft_data.f90
lfortran  -c utils.f90
lfortran  -c energies.f90
lfortran  -c ode1d.f90
lfortran  -c mixings.f90
lfortran  -c rdirac.f90
lfortran  -c rschroed.f90
lfortran  -c reigen.f90
lfortran  -c rpoisson.f90
lfortran  -c dft.f90
lfortran  -c mesh.f90
lfortran  -c states.f90
lfortran  -c drivers.f90
lfortran  -c dftatom.f90
ar rcs libdftatom.a constants.o dft.o dft_data.o dftatom.o drivers.o energies.o mesh.o mixings.o ode1d.o rdirac.o reigen.o rpoisson.o rschroed.o states.o types.o utils.o
warning: /Users/thirumalai/mambaforge/envs/fpm/bin/arm64-apple-darwin20.0.0-ranlib: archive library: libdftatom.a the table of contents is empty (no object file members in the library define global symbols)
cd tests; make -f Makefile.manual
cd lda; make -f Makefile.manual
lfortran  -I../../src -c conv_lda.f90
ASR verify pass error: ASR verify: The symbol table was not found in the scope of `symtab`.

ASR verify pass error: ASR verify: Struct::m_derived_type 'dft_data_t' cannot point outside of its symbol table, owner: r
Internal Compiler Error: Unhandled exception
Traceback (most recent call last):
  File "/Users/thirumalai/Open_Source/lfortran/src/bin/lfortran.cpp", line 2251
    return compile_to_object_file(arg_file, outfile, false,
  File "/Users/thirumalai/Open_Source/lfortran/src/bin/lfortran.cpp", line 962
    res = fe.get_llvm3(*asr, lpm, diagnostics, infile);
  File "/Users/thirumalai/Open_Source/lfortran/src/lfortran/fortran_evaluator.cpp", line 347
    compiler_options, run_fn, infile);
  File "/Users/thirumalai/Open_Source/lfortran/src/libasr/codegen/asr_to_llvm.cpp", line 8955
    pass_manager.apply_passes(al, &asr, pass_options, diagnostics);
  File "/Users/thirumalai/Open_Source/lfortran/src/libasr/pass/pass_manager.h", line 303
    apply_passes(al, asr, _passes, pass_options, diagnostics);
  File "/Users/thirumalai/Open_Source/lfortran/src/libasr/pass/pass_manager.h", line 168
    throw LCompilersException("Verify failed in the pass: "
LCompilersException: Verify failed in the pass: pass_array_by_data
make[2]: *** [conv_lda.o] Error 1
make[1]: *** [all] Error 2
make: *** [all] Error 2
```
This: https://github.com/czgdp1807/dftatom/blob/fe479fd27a7e0f77c6a88a1949996406ec935ac2/src/mixings.f90#L100 triggers the issue.

**Comment by Thirumalai-Shaktivel** [comment]
The branch: [czgdp1807/dftatom@lf8 (commits)](https://github.com/czgdp1807/dftatom/commits/lf8) compiles the file `conv_lda.f90 ` to LLVM (but has some LLVM error, we need to fix that)

**Comment by Thirumalai-Shaktivel** [comment]
I think, Next we have two files to compile, `conv_rlda.f90` and `F_atom_U.f90`.

**Comment by czgdp1807** [comment]
Okay. I will create MREs for these bugs. Seems like these are minor issues.

**Comment by czgdp1807** [comment]
MRE for LLVM error in https://github.com/lfortran/lfortran/issues/2386#issuecomment-1725895676,

```fortran
program debug
implicit none

type d
    real(8), dimension(:), pointer :: r
end type

real(8) :: array(10)
call f(array)

contains

subroutine f(array)
real(8), intent(out), target :: array(:)
real(8), target :: array1(size(array))
type(d) :: objd
objd%r => array1
end subroutine

end program
```

**Comment by Thirumalai-Shaktivel** [comment]
The seg-fault is fixed now!
Now, if I do: `make -f Makefile.manual test`, I get:
```console
$ make -f Makefile.manual test
cd tests; make -f Makefile.manual test
make[1]: Entering directory '/Users/thirumalai/Open_Source/dftatom/tests'
cd lda; make -f Makefile.manual test
make[2]: Entering directory '/Users/thirumalai/Open_Source/dftatom/tests/lda'
./conv_lda
Test eps: 1.19999999999999989e-03
3
3
Error in opening the file!
data.dat: No such file or directory
make[2]: *** [../Makefile.manual.mk:9: test] Error 1
make[2]: Leaving directory '/Users/thirumalai/Open_Source/dftatom/tests/lda'
make[1]: *** [Makefile.manual:12: test] Error 2
make[1]: Leaving directory '/Users/thirumalai/Open_Source/dftatom/tests'
make: *** [Makefile.manual:29: test] Error 2
```
There are bunch of files names data3.dat, data4.dat, ...
If I rename that files to data.dat, I get:
```console
$ make -f Makefile.manual test
cd tests; make -f Makefile.manual test
make[1]: Entering directory '/Users/thirumalai/Open_Source/dftatom/tests'
cd lda; make -f Makefile.manual test
make[2]: Entering directory '/Users/thirumalai/Open_Source/dftatom/tests/lda'
./conv_lda
Test eps: 1.19999999999999989e-03
3
3
Array sizes mismatch
STOP
make[2]: *** [../Makefile.manual.mk:9: test] Error 1
make[2]: Leaving directory '/Users/thirumalai/Open_Source/dftatom/tests/lda'
make[1]: *** [Makefile.manual:12: test] Error 2
make[1]: Leaving directory '/Users/thirumalai/Open_Source/dftatom/tests'
make: *** [Makefile.manual:29: test] Error 2
```

The same happens with GFortran as well.
Am I missing to execute some commands?

**Comment by Thirumalai-Shaktivel** [comment]
GFortran output:
```
./conv_lda
 Test eps:   1.1999999999999999E-003
 Z=          92
 N=         861
E_tot= -25658.4187880509 E_tot_exact= -25658.4178888534 error:  8.99E-04
 state    E            E_exact          error     occupancy
1s   -3689.3551425911  -3689.3551398369  2.75E-06    2.000
2s    -639.7787325283   -639.7787280866  4.44E-06    2.000
2p    -619.1085298771   -619.1085501807 -2.03E-05    6.000
3s    -161.1180902731   -161.1180732100  1.71E-05    2.000
3p    -150.9789760594   -150.9789801633 -4.10E-06    6.000
3d    -131.9773357481   -131.9773582831 -2.25E-05   10.000
4s     -40.5281007734    -40.5280842452  1.65E-05    2.000
4p     -35.8533242014    -35.8533208325  3.37E-06    6.000
4d     -27.1232017656    -27.1232122996 -1.05E-05   10.000
4f     -15.0274430071    -15.0274600691 -1.71E-05   14.000
5s      -8.8240983705     -8.8240894015  8.97E-06    2.000
5p      -7.0180949890     -7.0180922045  2.78E-06    6.000
5d      -3.8661716763     -3.8661751349 -3.46E-06   10.000
5f      -0.3665393509     -0.3665433531 -4.00E-06    3.000
6s      -1.3259789768     -1.3259763180  2.66E-06    2.000
6p      -0.8225386152     -0.8225379709  6.44E-07    6.000
6d      -0.1431896278     -0.1431901813 -5.53E-07    1.000
7s      -0.1309481392     -0.1309478622  2.77E-07    2.000
...
```

**Comment by certik** [comment]
TODO:
* Let's get all tests passing with workarounds
* Set it up at the CI
* Then each bugs is completely isolated and we can independently and in parallel fix all the remaining workarounds

One can do something like this:
```
diff --git a/tests/lda/conv_lda.f90 b/tests/lda/conv_lda.f90
index 0b34dc2..365b0ef 100644
--- a/tests/lda/conv_lda.f90
+++ b/tests/lda/conv_lda.f90
@@ -30,7 +30,7 @@ real(dp) :: eps
 real(dp), allocatable :: R(:), Rp(:), V_tot(:), density(:)
 integer :: p
 
-do p = 3, 8
+do p = 3, 5
     eps = 10.0_dp**(-p)
     eps = eps * 1.2_dp ! Allow numerical differences across compilers/platforms
     print *, "Test eps:", eps
```

**Comment by certik** [comment]
dftatom now compiles and works. If there are any remaining problems we can open dedicated issues for that.
